engine/crates/engine-v2/src/engine/trusted_documents.rs

```diff
@@ -0,0 +1,114 @@
+use super::{Engine, CLIENT_NAME_HEADER_NAME};
+use crate::response::GraphqlError;
+use engine::PersistedQueryRequestExtension;
+use runtime::trusted_documents::TrustedDocumentsError;
+use std::mem;
+
+impl Engine {
+    /// Handle a request making use of APQ or trusted documents.
+    pub(super) async fn handle_persisted_query(
+        &self,
+        request: &mut engine::Request,
+        client_name: Option<&str>,
+    ) -> Result<(), GraphqlError> {
+        let enforce_trusted_documents = self.env.trusted_documents.trusted_documents_enabled();
+        let persisted_query_extension = mem::take(&mut request.extensions.persisted_query);
+        let doc_id = mem::take(&mut request.operation_plan_cache_key.doc_id);
+
+        match (enforce_trusted_documents, persisted_query_extension, doc_id) {
+            (true, None, None) => Err(GraphqlError::new("Only trusted document queries are accepted.")),
```

Comment:
Maybe not the clearest message. We want to convey: this deployment is configured with trusted documents enabled. Only queries with a document id are allowed.

Reply:
Will improve.

Reply:
done :heavy_check_mark: